{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1",
   "metadata": {},
   "source": [
    "# ChEA-KG Appyter: Characterizing Enriched Transcription Factor Regulatory Subnetworks\n",
    "\n",
    "The [ChEA-KG](https://chea-kg.maayanlab.cloud/) Appyter predicts regulatory subnetworks of transcription factors (TFs) for an input gene set. The query gene sets are compared against libaries of TF target gene sets in ChEA3 to identify the most likely regulating TFs. The TFs are then connected via edges in the ChEA-KG background gene regulatory network (GRN). This network was constructed by submitting thousands of gene expression signatures from [RummaGEO](https://rummageo.com/) to [ChEA3](https://maayanlab.cloud/chea3/) for transcription factor enrichment analysis. The functions of enriched TFs are predicted using the [Gene Set Foundation Model](https://gsfm.maayanlab.cloud/). Automatic summaries of each subnetwork are produced using an LLM.\n",
    "\n",
    "The only required input is a gene list. Options to change other parameters such as the subnetwork size and layout are provided. It is recommended not to change the remaining parameters. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# general\n",
    "import requests\n",
    "import json\n",
    "import numpy as np \n",
    "import pandas as pd\n",
    "from IPython.display import display, HTML, Markdown\n",
    "import os\n",
    "import traceback\n",
    "\n",
    "# scatterplot\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from collections import OrderedDict\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "from bokeh.io import output_notebook\n",
    "from bokeh.plotting import figure, show\n",
    "from bokeh.models import HoverTool, ColumnDataSource\n",
    "import glasbey\n",
    "output_notebook()\n",
    "\n",
    "# bar chart\n",
    "import plotly.graph_objects as go\n",
    "import time\n",
    "import html\n",
    "\n",
    "# summary\n",
    "from google import genai\n",
    "from dotenv import load_dotenv\n",
    "load_dotenv()\n",
    "\n",
    "# subnetwork\n",
    "import ipycytoscape as ipc\n",
    "\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% do SectionField(\n",
    "    name='section1', \n",
    "    title = '1. Submit Your Gene List', \n",
    "    subtitle = 'Upload a text file containing your gene list -OR- copy and paste your gene list into the text box below (One gene per row). You can also try the default gene list provided.'\n",
    ") %}\n",
    "{% do SectionField(\n",
    "    name='section2', \n",
    "    title = '2. Describe your gene list', \n",
    "    subtitle = 'Provide a short description to describe your gene set.'\n",
    ") %}\n",
    "{% do SectionField(\n",
    "    name='section3', \n",
    "    title = '3. Choose subnetwork size', \n",
    "    subtitle = 'Choose the number of top-ranked enriched TFs included in each subnetwork. The maximum is 20 TFs. The default is 10.'\n",
    ") %}\n",
    "{% do SectionField(\n",
    "    name='section4', \n",
    "    title = '4. Choose subnetwork layout', \n",
    "    subtitle = 'Choose a Cytoscape-supported layout to view the network. The default is cose, a force-directed layout.'\n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% set gene_list_kind = TabField(\n",
    "    name='gene_list_kind',\n",
    "    label='Gene List',\n",
    "    default='Paste',\n",
    "    description='Paste or upload your gene list',\n",
    "    required=True,\n",
    "    choices={\n",
    "        'Paste': [\n",
    "            TextListField(\n",
    "                name='gene_list_input',\n",
    "                label='Gene List',\n",
    "                description='Paste your gene list (One gene per row):',\n",
    "                default=['MIR4454', 'RNU86', 'SNORD34', 'EEF1A1', 'RPL11', 'DCT', 'RPL37A', 'SNORD33', 'GAPDH', 'SNORD74', 'MIR4461', 'MIR4680', 'CD63', 'SNORD68', 'LDHA', 'TMSB4X', \n",
    "                'RPL27', 'SNORD108', 'MIR3191', 'RPS18', 'RPL41', 'ENO1', 'CAPG', 'RPS15A', 'SNORD79', 'FN1', 'LGALS3', 'GPNMB', 'NPM1', 'RPL7', 'RPS14', 'SNORD38A', 'RPS13', 'RPS7', \n",
    "                'ATOX1', 'PKM2', 'RPL31', 'SNORD76', 'SNORD42B', 'RPS29', 'BCYRN1', 'RPL6', 'ATP5E', 'RPS3A', 'RPL39', 'YWHAZ', 'TOMM7', 'RPS27A', 'SNORD49A', 'MIR4482-1', 'SNORD5', \n",
    "                'RPL30', 'MIR1292', 'RPL5', 'SNORD59A', 'RPS21', 'PSAP', 'RPL35A', 'RPL13AP5', 'SNORD50B', 'H2AFZ', 'SNORD27', 'PPIA', 'PRDX1', 'RPL21P28', 'RPL9', 'RPS12', 'HSP90B1', \n",
    "                'COX7C', 'RPL23', 'SNORD30', 'LDHB', 'GNB2L1', 'SGK1', 'FKBP1A', 'SNORD57', 'HSPD1', 'SNORD12', 'MIR4273', 'RPS4X', 'UQCRH', 'RPS2', 'COX5B', 'ATP5B', 'MIR4687', \n",
    "                'MIR4263', 'RPS10', 'RAN', 'MIR3687', 'PTTG1', 'SNORD87', 'ATP1A1', 'NME2', 'SNORD18A', 'RPSA', 'TUBB', 'RPL22', 'RPLP0', 'SNORD101', 'ANXA5', 'CD74', 'RPS6', \n",
    "                'MIR4691', 'GNAS', 'GSTO1', 'EIF4A1', 'NQO1', 'SNAR-G1', 'MIR4653', 'UBB', 'RPL38', 'SNORD4A', 'SNORD82', 'MIA', 'SNORD37', 'EIF3E', 'MIR4678', 'PDIA6', 'SLC25A3', \n",
    "                'PARK7', 'PGK1', 'C17orf76-AS1', 'YBX1', 'NCL', 'RPL35', 'HSP90AB1', 'MIR103B2', 'RPL26', 'DSTN', 'SNHG8', 'CALR', 'EEF1G', 'ATP5J2', 'TUBA1B', 'CSTB', 'SPP1', 'CALU', \n",
    "                'PABPC1', 'PRAME', 'LY6E', 'HNRNPA2B1', 'SNORD38B', 'SLC25A5', 'FABP5', 'MIR1915', 'SERPINF1', 'RPS8', 'DDOST', 'HINT1', 'RPL18A', 'RPS20', 'PSMA7', 'CHCHD2', 'HNRNPA1',\n",
    "                 'RPL15', 'PSMA1', 'HLA-DRA', 'RPS3', 'RPL27A', 'EEF1B2', 'TXN', 'RPL4', 'SNORD16', 'PRNP', 'MDH1', 'NME1', 'CANX', 'SNORD35B', 'TBCA', 'TPI1', 'LOC645591', 'PTMA', \n",
    "                 'ATP5A1', 'CBX3', 'SDCBP', 'C1QBP', 'DBI', 'SNORD59B', 'PLA1A', 'RPL37', 'NACA', 'CDK2', 'MIR324', 'RPS9', 'ALDOA', 'COX6A1', 'RPN2', 'ATP5F1', 'ATP5G3', 'SNAR-E', \n",
    "                 'GMPR', 'SNRPD2', 'MIR4700', 'GYPC', 'CTSK', 'SHFM1', 'P4HB', 'CTSZ', 'MIR1260B', 'PCNA', 'HMGB1', 'COX7B', 'TM4SF1', 'SNORA41', 'CTSB', 'SLC20A1', 'IER3', 'ACSL3', \n",
    "                 'CKS2', 'ATP1B3', 'MIR3188', 'ZNFX1-AS1', 'COX7A2', 'MIR135A1', 'PTGES3', 'CSDE1', 'LAMP2', 'SDC3', 'AMD1', 'MIR4523', 'SPON2', 'RPL10A', 'MIR4639', 'MIR4517', 'XRCC6', \n",
    "                 'CSE1L', 'RPL8', 'MIR3653', 'MIR3190', 'SNORD54', 'RPL14', 'NBL1', 'ACTR3', 'ATP5C1', 'SNORD22', 'ATP1A1OS', 'EIF5A', 'SNORA10', 'GNG12', 'EIF3L', 'YWHAE', 'VDAC1', \n",
    "                 'CD109', 'SRP9', 'ATP5H', 'SLIRP', 'MFI2', 'RPL19', 'LPXN', 'CLIC4', 'BTF3', 'HLA-A', 'FAM167B', 'PDIA3', 'SEC61G', 'MGST1', 'TXNRD1', 'ATP5G1', 'LITAF', 'HSD17B12', \n",
    "                 'IVNS1ABP', 'SNORD31', 'NT5E', 'A2M', 'UBA52', 'POMP', 'HMGN2', 'ARL6IP5', 'HSP90AA1', 'SIRPA', 'EMP3', 'WFDC1', 'EIF3D', 'PYGB', 'SSBP1', 'SNRPB', 'RPL32', 'CCT7', \n",
    "                 'IFI6', 'MCAM', 'RPL10', 'XRCC5', 'ATP6V1E1', 'SNRPG', 'MITF', 'RPL13A', 'MIR2861', 'C3orf14', 'C14orf2', 'RPS5', 'FBXO7', 'SPARC', 'SYPL1', 'RGS10', 'SLC45A2', 'APP', \n",
    "                 'ANXA1', 'CD68', 'CCT2', 'IPO7', 'CCT4', 'HNRNPA3', 'CAP1', 'HSPE1', 'MBP', 'ACTR2', 'UCN2', 'MIR25', 'CASP1', 'EIF3I', 'SMS', 'MME', 'ARPC2', 'CDC42', 'NDUFB9', \n",
    "                 'AP1S2', 'PRDX3', 'SRPX', 'PHGDH', 'FBL', 'CTSC', 'SNORA20', 'HNRNPH1', 'PDIA4', 'EIF3H', 'SOAT1', 'VGF', 'GANAB', 'HSPA9', 'GLO1', 'PRKAR1A', 'SNRPF', 'SDHB', 'TIMP1', \n",
    "                 'PSMD6', 'BCL2A1', 'SNRPB2', 'NDUFB3', 'SNHG6', 'CORO1C', 'THOC7', 'SNX10', 'CEACAM1', 'LAPTM4B', 'SFRP1', 'ARPC3', 'G3BP1', 'COX17', 'GPM6B', 'SSR3', 'ETFB', 'MIR4665', \n",
    "                 'CCT8', 'SLC43A3', 'GJB1', 'EIF4B', 'RPL18', 'KPNA2', 'CAPZB', 'FABP7', 'NOP56', 'HNRNPK', 'ERP29', 'VAMP8', 'OAT', 'PSMB1', 'CTSH', 'NSA2', 'LGALS3BP', 'SSB', 'LUZP6', \n",
    "                 'POLE4', 'TIMP3', 'SF3B14', 'SUMO1', 'UGP2', 'PSMA2', 'PEG10', 'ERGIC3', 'ERH', 'MIR4785', 'C19orf79', 'PLP2', 'AKR1B1', 'AZIN1', 'RAB38', 'ADSL'],\n",
    "                section='section1'\n",
    "            ),\n",
    "        ],\n",
    "        'Upload': [\n",
    "            FileField(\n",
    "                name='gene_list_filename',\n",
    "                label='Gene List File',\n",
    "                default='',\n",
    "                description='Upload your gene list as a text file (One gene per row).',\n",
    "                section='section1'\n",
    "            ),\n",
    "        ],\n",
    "    },\n",
    "    section = 'section1',\n",
    ") %}\n",
    "{%\n",
    "    set description = TextField(\n",
    "        name='description',\n",
    "        label=\"Description\",\n",
    "        default=\"Genes upregulated in the SKMEL28 melanoma cell line vs. primary melanocytes (GSE88741)\",\n",
    "        description=\"Provide a short description of your gene set\",\n",
    "        section = 'section2',\n",
    "    )\n",
    "%}\n",
    "{%\n",
    "    set limit = IntField(\n",
    "        name='subnetwork-size',\n",
    "        label=\"Subnetwork Size\",\n",
    "        default=10,\n",
    "        min = 1,\n",
    "        max = 20,\n",
    "        description=\"Choose the subnetwork size\",\n",
    "        section = 'section3',\n",
    "    )\n",
    "%}\n",
    "{%\n",
    "    set layout = ChoiceField(\n",
    "        name='subnetwork-layout',\n",
    "        label=\"Subnetwork Layout\",\n",
    "        default=\"cose\",\n",
    "        description=\"Choose the subnetwork layout. The default is 'cose', a force-directed layout. Learn more about Cytoscape layouts here: https://blog.js.cytoscape.org/2020/05/11/layouts/\",\n",
    "        section = 'section4',\n",
    "        choices = [\n",
    "            \"cose\",\n",
    "            \"grid\",\n",
    "            \"circle\",\n",
    "            \"breadthfirst\"\n",
    "        ]\n",
    "    )\n",
    "%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{%- if gene_list_kind.raw_value == 'Paste' %}\n",
    "gene_list_input = {{ gene_list_kind.value[0] }}\n",
    "{%- else %}\n",
    "gene_list_filename = {{ gene_list_kind.value[0] }}\n",
    "{%- endif %}\n",
    "\n",
    "{%- if gene_list_kind.raw_value == 'Paste' %}\n",
    "genes = [x.strip() for x in gene_list_input]\n",
    "{%- else %}\n",
    "open_gene_list_file = open(gene_list_filename,'r')\n",
    "lines = open_gene_list_file.readlines()\n",
    "genes = [x.strip() for x in lines]\n",
    "open_gene_list_file.close()\n",
    "{%- endif %}\n",
    "\n",
    "term_limit = {{limit}}\n",
    "threshold = 3 #used for ChEA3 to determine min number of libraries for meanrank\n",
    "sn_layout = \"{{layout.value}}\"\n",
    "desc = {{description}}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6",
   "metadata": {},
   "source": [
    "## Extracting enriched subnetwork from ChEA-KG\n",
    "First, we extract the enriched subnetwork from ChEA-KG programmatically. This requires two steps:\n",
    "\n",
    "**1. Add the gene list to the ChEA-KG site, where it will be accessbile via a unique ID (`userListId`). <ins>This ID can be used to share results but is private by default.</ins>**  \n",
    "*Input:* List of newline separated genes and a description (optional)    \n",
    "*Returns:* `userListId`, the unique ID associated with that gene set   \n",
    "\n",
    "**2. Retrieve enrichment results using the gene list ID**   \n",
    "*Input:* `userListId`, `min_lib` and `libraries` parameters. ChEA-KG enrichment uses the MeanRank method from ChEA3 [(PMID31114921)](https://academic.oup.com/nar/article/47/W1/W212/5494769?login=false), which takes the average rank of a TF across all 6 primary libraries. `min_lib` requires by default that a TF is ranked in at least 3 libraries but can take any value from 1-6. `term_limit` specifies the number of top-ranked TFs that are returned    \n",
    "*Output:* JSON formatted list of enriched nodes and the edges  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cheakg_results(chea_gene_list, desc=\"\"):\n",
    "    '''\n",
    "    Find the subnetwork of enriched TFs for an input gene list\n",
    "    '''\n",
    "    CHEA_KG = 'https://chea-kg.maayanlab.cloud/api/enrichment'\n",
    "        \n",
    "    payload = {\n",
    "        'list': (None, \"\\n\".join(chea_gene_list)),\n",
    "        'description': (None, desc)\n",
    "    }\n",
    "    try:\n",
    "        response=requests.post(f\"{CHEA_KG}/addList\", files=payload)\n",
    "        data = json.loads(response.text)\n",
    "    except Exception as e: \n",
    "        print(\"Error connecting to ChEA-KG: \", e)\n",
    "    \n",
    "    q = {\n",
    "        'min_lib': 3, # minimum number of libraries that a TF must be ranked in\n",
    "        'libraries': [\n",
    "            {'library': \"Integrated--meanRank\", 'term_limit': term_limit} # edit term_limit to change number of top-ranked TFs\n",
    "        ],\n",
    "        'limit':50, # controls number of edges returned - may cause issues with visualization if too large\n",
    "        'userListId': data['userListId']\n",
    "    }\n",
    "    \n",
    "    query_json=json.dumps(q)\n",
    "    \n",
    "    res = requests.post(CHEA_KG, data=query_json)\n",
    "    if res.ok:\n",
    "        data = json.loads(res.text)\n",
    "    else:\n",
    "        data = None\n",
    "        print(res.text)\n",
    "    return data\n",
    "\n",
    "def extract_nodes(cytoscape_json):\n",
    "    '''\n",
    "    Extract list of node labels and IDs from a cytoscape object\n",
    "    '''\n",
    "    nodes = []\n",
    "    for entry in cytoscape_json['nodes']:\n",
    "        nodes.append({'id':entry['data']['id'], 'label':entry['data']['label']})\n",
    "    return nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "d = get_cheakg_results(genes, desc)\n",
    "nodes = extract_nodes(d)\n",
    "nodelist = [x['label'] for x in nodes]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9",
   "metadata": {},
   "source": [
    "## UMAP of ChEA-KG TFs\n",
    "Each point in the UMAP represents a TF with at least one source relationship (n=700). Term frequency-inverse document frequency (TF-IDF) values were computed for the target TF set corresponding to each source TF, and UMAP was applied to the resulting values. The TFs are plotted based on the first two UMAP dimensions. Generally, TFs with more similar target TF sets are positioned closer together. TFs are colored by automatically identified clusters computed with the Leiden algorithm applied to the TF-IDF values. Hovering over points will display the TF and the automatically assigned cluster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_scatterplot(libdict, nneighbors=30, mindist=0.1, spread=1.0, maxdf=1.0, mindf=1):\n",
    "    vec = TfidfVectorizer(max_df=maxdf, min_df=mindf)\n",
    "    X = vec.fit_transform(libdict.values())\n",
    "    adata = anndata.AnnData(X)\n",
    "    adata.obs.index = libdict.keys()\n",
    "\n",
    "    ### the n_neighbors and min_dist parameters can be altered\n",
    "    sc.pp.neighbors(adata, n_neighbors=nneighbors)\n",
    "    sc.tl.leiden(adata, resolution=1.0)\n",
    "    sc.tl.umap(adata, min_dist=mindist, spread=spread, random_state=42)\n",
    "\n",
    "    new_order = adata.obs.sort_values(by='leiden').index.tolist()\n",
    "    adata = adata[new_order, :]\n",
    "    adata.obs['leiden'] = 'Cluster ' + adata.obs['leiden'].astype('object')\n",
    "\n",
    "    df = pd.DataFrame(adata.obsm['X_umap'])\n",
    "    df.columns = ['x', 'y']\n",
    "\n",
    "    df['cluster'] = adata.obs['leiden'].values\n",
    "    df['term'] = adata.obs.index\n",
    "    df['genes'] = [libdict[l] for l in df['term']]\n",
    "\n",
    "    return df\n",
    "\n",
    "def get_scatter_colors(df):\n",
    "    clusters = pd.unique(df['cluster']).tolist()\n",
    "    colors = glasbey.create_palette(palette_size=len(clusters), lightness_bounds=(0,100), chroma_bounds=(50,100), as_hex=True)\n",
    "    color_mapper = {clusters[i]: colors[i % 20] for i in range(len(clusters))}\n",
    "    return color_mapper\n",
    "\n",
    "def get_scatterplot(scatterdf, nodes):\n",
    "    df = scatterdf.copy()\n",
    "    color_mapper = get_scatter_colors(df)\n",
    "    df['color'] = df['cluster'].apply(lambda x: color_mapper[x])\n",
    "    df['lw'] = df['term'].apply(lambda x: 2 if x in nodes else 0.5)\n",
    "    df['la'] = df['term'].apply(lambda x: 1 if x in nodes else 0)\n",
    "    df['size'] = df['term'].apply(lambda x: 12 if x in nodes else 5)\n",
    "    df['fa'] = df['term'].apply(lambda x: 1 if x in nodes else 0.5)\n",
    "\n",
    "    tooltips = [\n",
    "        (\"Gene Set\", \"@gene_set\"),\n",
    "        (\"Cluster\", \"@label\")\n",
    "    ]\n",
    "        \n",
    "    hover_emb = HoverTool(tooltips=tooltips)\n",
    "    tools_emb = [hover_emb, 'pan', 'wheel_zoom', 'reset', 'save']\n",
    "\n",
    "    plot_emb = figure(\n",
    "        width=900, \n",
    "        height=700, \n",
    "        tools=tools_emb\n",
    "    )\n",
    "\n",
    "    source = ColumnDataSource(\n",
    "        data=dict(\n",
    "            x = df['x'],\n",
    "            y = df['y'],\n",
    "            gene_set = df['term'],\n",
    "            sizes = df['size'],\n",
    "            colors = df['color'],\n",
    "            fill_alphas = df['fa'],\n",
    "            label = df['cluster'],\n",
    "            line_widths = df['lw'],\n",
    "            line_alphas = df['la'],\n",
    "            line_colors = ['#000000']*df.shape[0]\n",
    "        )\n",
    "    )\n",
    "\n",
    "    # hide axis labels and grid lines\n",
    "    plot_emb.xaxis.major_tick_line_color = None\n",
    "    plot_emb.xaxis.minor_tick_line_color = None\n",
    "    plot_emb.yaxis.major_tick_line_color = None\n",
    "    plot_emb.yaxis.minor_tick_line_color = None\n",
    "    plot_emb.xaxis.major_label_text_font_size = '0pt'\n",
    "    plot_emb.yaxis.major_label_text_font_size = '0pt' \n",
    "\n",
    "    plot_emb.output_backend = \"svg\"    \n",
    "    \n",
    "    plot_emb.xaxis.axis_label = \"UMAP_1\"\n",
    "    plot_emb.yaxis.axis_label = \"UMAP_2\"\n",
    "    \n",
    "    s = plot_emb.scatter(\n",
    "            'x', \n",
    "            'y', \n",
    "            size = 'sizes', \n",
    "            source = source, \n",
    "            color = 'colors', \n",
    "            fill_alpha = 'fill_alphas',\n",
    "            line_alpha = 'line_alphas',\n",
    "            line_width = 'line_widths',\n",
    "            line_color = 'line_colors',\n",
    "            legend_group = 'label'\n",
    "        )\n",
    "\n",
    "    plot_emb.add_layout(plot_emb.legend[0], 'right')\n",
    "    \n",
    "    return plot_emb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": [
    "r = requests.get(\"https://minio.dev.maayanlab.cloud/hgrn-chear/network_target_sets.gmt\")\n",
    "file = r.text.split(\"\\n\")\n",
    "\n",
    "### variables to store gene set data\n",
    "lib_dict = OrderedDict()\n",
    "\n",
    "for line in file[:-1]:\n",
    "    tokens = line.split(\"\\t\\t\")\n",
    "    term = tokens[0]\n",
    "    genes = [x.split(',')[0].strip() for x in tokens[1].split('\\t')]\n",
    "    lib_dict[term] = ' '.join(genes)\n",
    "\n",
    "## defaults: nneighbors=30, mindist=0.1, spread=1.0, maxdf=1.0, mindf=1\n",
    "scatter_df = process_scatterplot(\n",
    "    lib_dict, \n",
    "    nneighbors=5,\n",
    "    mindist=.05,\n",
    ")\n",
    "\n",
    "plot = get_scatterplot(scatter_df, nodelist)\n",
    "show(plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12",
   "metadata": {},
   "source": [
    "**Figure 1.** *UMAP embedding of ChEA-KG source TFs. Enriched TFs for the input gene set are enlarged and outlined. Each point represents a TF with at least one source relationship. Points are colored based on their leiden cluster membership.*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13",
   "metadata": {},
   "source": [
    "## Bar Chart View of Enriched TFs\n",
    "Enriched TFs are identified using the [ChEA3 api](https://maayanlab.cloud/chea3/). The bar chart below provides a more detailed view of how the TF rankings are calculated. An enrichment score is calculated for each TF against each of the six ChEA3 background libraries of TF-target sets. The TFs are ranked for each library based on their enrichment scores. The final TF rankings are based on the MeanRank, which is the average rank of each TF across all libraries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_chea3_results(gene_set, query_name):\n",
    "    ADDLIST_URL = 'https://maayanlab.cloud/chea3/api/enrich/'\n",
    "    payload = {\n",
    "        'gene_set': gene_set,\n",
    "        'query_name': query_name\n",
    "    }\n",
    "    response = requests.post(ADDLIST_URL, data=json.dumps(payload))\n",
    "    if not response.ok: \n",
    "        # r.ok (where r is the object) returns whether the call to the url was successful\n",
    "        raise Exception('Error analyzing gene list')\n",
    "    time.sleep(1)\n",
    "    return json.loads(response.text) # .text returns the content of response in unicode\n",
    "\n",
    "# Function for displaying the individual library bar charts \n",
    "def display_charts(libs, description): \n",
    "    for libname in libs:\n",
    "        \n",
    "        display(HTML(f'<h3>{libname}</h3>'))\n",
    "        \n",
    "        tfs = [i['TF'] for i in results[libname]][0:term_limit]\n",
    "        scores = [float(i['FET p-value']) for i in results[libname]][0:term_limit]\n",
    "        \n",
    "        # reverse the order/ranking of the tfs (and their respective scores)\n",
    "        tfs = tfs[::-1]\n",
    "        scores = scores[::-1]\n",
    "\n",
    "        # takes the -log of the scores\n",
    "        scores = -np.log10(scores)\n",
    "\n",
    "        \n",
    "        score_range = max(scores) - min(scores)\n",
    "        x_lowerbound = min(scores) - (score_range * 0.05)\n",
    "        x_upperbound = max(scores) + (score_range * 0.05)\n",
    "        \n",
    "        libfig = go.Figure(data = go.Bar(name = libname, \n",
    "                                         x = scores, \n",
    "                                         y = tfs, \n",
    "                                         marker = go.bar.Marker(color = 'rgb(255,127,80)'), \n",
    "                                         orientation = 'h'))\n",
    "        libfig.update_layout(\n",
    "            title = {\n",
    "                'text':'Bar Chart of Scores based on FET p-values',\n",
    "                'y': 0.87,\n",
    "                'x': 0.5,\n",
    "                'xanchor':'center',\n",
    "                'yanchor':'top'\n",
    "            },\n",
    "            xaxis_title = '-log\\u2081\\u2080(FET p-value)', \n",
    "            # \\u208 unicode to get the subscript (need a subscript of \"10\")\n",
    "            yaxis_title = 'Transcription Factors',\n",
    "            font = dict(\n",
    "                size = 16,\n",
    "                color = 'black'\n",
    "            )\n",
    "        )\n",
    "        \n",
    "        libfig.update_xaxes(range = [x_lowerbound, x_upperbound])\n",
    "        \n",
    "        libfig.show()\n",
    "        \n",
    "        display(HTML(f'<h5>{description[libname]}</h5>'))\n",
    "        \n",
    "def indexfinder(lib_score_list, value):\n",
    "    index = 1\n",
    "    for num in lib_score_list:\n",
    "        if num == value:\n",
    "            return index\n",
    "        elif num != 0:\n",
    "            index += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15",
   "metadata": {},
   "outputs": [],
   "source": [
    "results = get_chea3_results(genes, 'query')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16",
   "metadata": {},
   "outputs": [],
   "source": [
    "c_lib_palette = {'ARCHS4 Coexpression':'rgb(196, 8, 8)',\n",
    "                 'ENCODE ChIP-seq':'rgb(244, 109, 67)',\n",
    "                 'Enrichr Queries':'rgb(242, 172, 68)', \n",
    "                 'GTEx Coexpression':'rgb(236, 252, 68)',\n",
    "                 'Literature ChIP-seq':'rgb(165, 242, 162)',\n",
    "                 'ReMap ChIP-seq':'rgb(92, 217, 78)'}\n",
    "# this sets all the color values for all the libraries that will be displayed in the bar chart\n",
    "\n",
    "# NOTE: removed Integrated mean/topRank since those are compiled from the above 6 libraries \n",
    "# afterwards and so none of the TFs will have Integrated mean/topRank as one of their libraries\n",
    "\n",
    "c_lib_means = {'ARCHS4 Coexpression': [0] * term_limit, 'ENCODE ChIP-seq': [0] * term_limit, \n",
    "               'Enrichr Queries': [0] * term_limit, 'GTEx Coexpression': [0] * term_limit,\n",
    "               'Literature ChIP-seq': [0] * term_limit, 'ReMap ChIP-seq': [0] * term_limit}\n",
    "# creates a dictionary where each library is a key, and the values are empty lists with as\n",
    "# many indices/spaces as the user has requested transcription factors (ex: if the user\n",
    "# requests 15 TFs to be returned, the lists will have 15 spaces)\n",
    "\n",
    "\n",
    "libs_sorted = ['ARCHS4 Coexpression','ENCODE ChIP-seq','Enrichr Queries',\n",
    "               'GTEx Coexpression','Literature ChIP-seq','ReMap ChIP-seq']\n",
    "\n",
    "\n",
    "mr_results = results['Integrated--meanRank']\n",
    "###### NOTE: for meanRank, the TFs are already ranked by Score ######\n",
    "\n",
    "for i in range(len(mr_results)):\n",
    "    for lib in libs_sorted:\n",
    "        mr_results[i].update({lib:0})\n",
    "        \n",
    "for i in range(len(mr_results)):\n",
    "    thing = mr_results[i]['Library'].split(';')\n",
    "    for a in range(len(thing)):\n",
    "        library, value = thing[a].split(',')\n",
    "        mr_results[i].update({library:int(value)})\n",
    "    \n",
    "sortedARCHS4 = sorted(mr_results, key = lambda k: k['ARCHS4 Coexpression'])\n",
    "sortedGTEx = sorted(mr_results, key = lambda k: k['GTEx Coexpression']) \n",
    "sortedEnrichr = sorted(mr_results, key = lambda k: k['Enrichr Queries']) \n",
    "sortedENCODE = sorted(mr_results, key = lambda k: k['ENCODE ChIP-seq']) \n",
    "sortedReMap = sorted(mr_results, key = lambda k: k['ReMap ChIP-seq']) \n",
    "sortedLit = sorted(mr_results, key = lambda k: k['Literature ChIP-seq']) \n",
    "\n",
    "rankedARCHS4 = [entry['ARCHS4 Coexpression'] for entry in sortedARCHS4]\n",
    "rankedENCODE = [entry['ENCODE ChIP-seq'] for entry in sortedENCODE]\n",
    "rankedEnrichr = [entry['Enrichr Queries'] for entry in sortedEnrichr] \n",
    "rankedGTEx = [entry['GTEx Coexpression'] for entry in sortedGTEx]\n",
    "rankedLit = [entry['Literature ChIP-seq'] for entry in sortedLit]\n",
    "rankedReMap = [entry['ReMap ChIP-seq'] for entry in sortedReMap] \n",
    "\n",
    "\n",
    "ranking_dict = {'ARCHS4 Coexpression':rankedARCHS4,\n",
    "                'ENCODE ChIP-seq':rankedENCODE,\n",
    "                'Enrichr Queries':rankedEnrichr,\n",
    "                'GTEx Coexpression':rankedGTEx,\n",
    "                'Literature ChIP-seq':rankedLit,\n",
    "                'ReMap ChIP-seq':rankedReMap}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17",
   "metadata": {},
   "outputs": [],
   "source": [
    "for tfentry in mr_results:\n",
    "    tfentry.update( [('SumRank', 0), ('AvgRank', 0) ])\n",
    "    library_scores = tfentry['Library'].split(';')\n",
    "    lib_counter = 0\n",
    "    for a in library_scores:\n",
    "        l, v = a.split(',')\n",
    "        v = int(v)\n",
    "        #scorerank = ranking_dict[l].index(v) + 1\n",
    "        scorerank = indexfinder(ranking_dict[l], int(v))\n",
    "        tfentry['SumRank'] += int(scorerank)\n",
    "        lib_counter += 1\n",
    "    tfentry['AvgRank'] = (tfentry['SumRank'] / lib_counter)\n",
    "    \n",
    "sorted_results = sorted(mr_results, key = lambda k: k['AvgRank'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18",
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted_top_results = []\n",
    "index = 0\n",
    "while (len(sorted_top_results) < term_limit):\n",
    "    if len(sorted_results[index]['Library'].split(';')) >= threshold:\n",
    "        sorted_top_results.append(sorted_results[index])\n",
    "    index += 1\n",
    "    # moves on to the next index\n",
    "    \n",
    "sorted_top_results = sorted_top_results[::-1]\n",
    "\n",
    "# set up a list with all the TFs, sorted by rank (lowest to highest, in line with top_results)\n",
    "sorted_tfs = []\n",
    "for i in range(0, len(sorted_top_results)):\n",
    "    sorted_tfs.append(sorted_top_results[i].get('TF'))\n",
    "    # this pulls only the TF name from top_results and adds it to sorted_tfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, tfentry in enumerate(sorted_top_results):\n",
    "    libscores = tfentry['Library'].split(';')\n",
    "    for a in libscores:\n",
    "        lib, value = a.split(',')\n",
    "        rank = indexfinder(ranking_dict[lib], int(value))\n",
    "        avg = tfentry['AvgRank']\n",
    "        tot = tfentry['SumRank']\n",
    "        bar_length = (rank*avg)/tot\n",
    "        c_lib_means[lib][i] = float(bar_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting the actual bar chart\n",
    "fig = go.Figure(data = [go.Bar(name = c_lib, \n",
    "                               x = c_lib_means[c_lib], \n",
    "                               y = sorted_tfs,\n",
    "                               marker = go.bar.Marker(color = c_lib_palette[c_lib]), \n",
    "                               orientation = 'h') \n",
    "                        for c_lib in libs_sorted])\n",
    "h = 400 if term_limit <=10 else 400+10*term_limit\n",
    "fig.update_layout(barmode = 'stack')\n",
    "'''fig.update_layout(\n",
    "    title = {\n",
    "        'text': 'Stacked Bar Chart of Average Ranks in Different Libraries',\n",
    "        'y': 0.67,\n",
    "        'x': 0.5,\n",
    "        'xanchor': 'center',\n",
    "        'yanchor': 'top',\n",
    "    }\n",
    ")'''\n",
    "fig.update_layout(\n",
    "    \n",
    "    xaxis_title = 'Average of Ranks Across All Libraries',\n",
    "    yaxis_title = 'Transcription Factors',\n",
    "    font = dict(\n",
    "        size = 16,\n",
    "        color = 'black'\n",
    "    ),\n",
    "    width=900,\n",
    "    height=h\n",
    ")\n",
    "html_str = fig.to_html(include_plotlyjs='cdn')\n",
    "escaped_html = html.escape(html_str)\n",
    "\n",
    "iframe = f\"\"\"\n",
    "<iframe srcdoc=\"{escaped_html}\" width=\"100%\" height=\"{h}\" frameborder=\"0\"></iframe>\n",
    "\"\"\"\n",
    "\n",
    "display(HTML(iframe))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21",
   "metadata": {},
   "source": [
    "**Figure 2.** *Bar chart of enriched transcription factors. The bar chart displays the top ranked transcription factors based on their average rank across each of the TF-target libraries used in the ChEA3 enrichment calculation. Each bar displays a top ranked TF along with its average score for each library.* "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22",
   "metadata": {},
   "source": [
    "## Automatically Generated Summary of Subnetwork TFs\n",
    "The below summary is automatically generated using Google's [Gemma 3n E4B](https://huggingface.co/google/gemma-3n-E4B-it-litert-preview) model. Interpret results with caution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:    \n",
    "    API_KEY = os.environ['CHEAKG_GEMINI_API_KEY']\n",
    "    client = genai.Client(api_key=API_KEY)\n",
    "    MODEL = 'gemma-3n-e4b-it'\n",
    "\n",
    "    prompt = f'You are a biomedical researcher who is investigating transcriptional regulatory subnetworks. \\\n",
    "        You are writing a summary of the top {term_limit} enriched transcription factors for an input gene set, predicted using ChEA3. The transcription factors are {nodelist}.\\\n",
    "        Write five sentences summarizing the biological function of this subnetwork. Include AMA style citations from pubmed. Your reponse should only contain the five \\\n",
    "            sentences and references. Separate the references section from the rest of your response using an empty line.\\\n",
    "            '\n",
    "\n",
    "    response = client.models.generate_content(\n",
    "        model=MODEL, contents=prompt\n",
    "    )\n",
    "\n",
    "    res = response.text.replace(r\"(\\r?\\n){2,}\", \"\").replace(\"'\",\"\")#.replace(\"] \",\"]\")\n",
    "    display(Markdown(res))\n",
    "except:\n",
    "    print(\"An API key is required to use this feature. API keys can be generated in Google AI Studio: https://aistudio.google.com/apikey. The model used in this Appyter (Gemma 3n E4B) is open-source and free to use.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24",
   "metadata": {},
   "source": [
    "## Subnetwork Phenotype predictions with KOMP2\n",
    "The Gene Set Foundation Model (GSFM) [(PMID:40501705)](https://pubmed.ncbi.nlm.nih.gov/40501705/) is trained to recover held-out genes from gene sets. Below, we use this tool to make predictions about the gene set membership of subnetwork TFs using the [Knockout Mouse Phenotyping Program](https://www.mousephenotype.org/) (KOMP2) Mouse Phenotypes 2022 [gene set library](https://maayanlab.cloud/Enrichr/#libraries). We report the top 10 phenotype predictions across all TFs in the network. \n",
    "\n",
    "To generate consensus TF predictions, we predict terms individually for each enriched TF and select the top 10 from the set of all predictions, ranked by score and z-score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = 'rummagene'\n",
    "SOURCE = 'KOMP2_Mouse_Phenotypes_2022'\n",
    "cols = ['term','score','zscore','gene']\n",
    "\n",
    "predict_df = pd.DataFrame()\n",
    "for n in nodes:\n",
    "    gene = n['label']\n",
    "    \n",
    "    # extract sources\n",
    "    try:\n",
    "      res = requests.get('https://gsfm.maayanlab.cloud/api/trpc/sources', params={\n",
    "        'input': json.dumps({\"model\":model,\"gene\":gene})\n",
    "      }).json()\n",
    "\n",
    "      # extract predictions\n",
    "    except Exception as e:\n",
    "      print(f\"Error retrieving sources: {e}\")\n",
    "\n",
    "    sources = { row['source'] for row in res['result']['data'] }\n",
    "\n",
    "    try:\n",
    "        res = requests.get(\n",
    "          f\"https://gsfm.maayanlab.cloud/api/trpc/{','.join('predictions' for source in sources)}\", params={\n",
    "            'batch': '1',\n",
    "            'input': json.dumps({\n",
    "              str(i): {\"model\":model,\"source\":source,\"gene\":gene,\"offset\":0,\"limit\":10}\n",
    "              for i, source in enumerate(sources)\n",
    "            }),\n",
    "          }\n",
    "        ).json() \n",
    "    except Exception as e:\n",
    "        print(f\"Error retrieving predictions: {e}\")\n",
    "        traceback.print_exc()\n",
    "\n",
    "    all_predictions = dict(zip(sources, [row['result']['data'] for row in res]))\n",
    "    predictions = all_predictions.get(SOURCE, [])\n",
    "    entries = [(v['term'], round(v['proba'], 2), v['zscore'], gene) for v in predictions]\n",
    "    temp = pd.DataFrame(entries, columns=cols)\n",
    "    if not temp.empty:\n",
    "      predict_df = pd.concat([predict_df,temp])\n",
    "    time.sleep(0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26",
   "metadata": {},
   "outputs": [],
   "source": [
    "consensus_predictions = predict_df.sort_values(by=['score','zscore'], ascending = False)[0:10].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27",
   "metadata": {},
   "outputs": [],
   "source": [
    "display(HTML(consensus_predictions.to_html(index=False)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28",
   "metadata": {},
   "source": [
    "**Table 1.** *Consensus TF predictions from GSFM based on the KOMP2 Mouse Phenotypes 2022 gene set library. The table shows the top 10 highest scoring predictions for all enriched TFs, along with the model-assigned score and z-score for each term. The \"gene\" column indicates the TF for which the high-scoring function prediction was made.*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29",
   "metadata": {},
   "source": [
    "## Visualizing Enriched Subnetwork with Cytoscape\n",
    "The ChEA-KG network visualization is built using [Cytoscape.js](https://js.cytoscape.org/). We can recreate this visualization in Python using the `ipycytoscape` [widget](https://github.com/cytoscape/ipycytoscape). Be careful of zooming out too far - it can be challenging to find the graph again!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30",
   "metadata": {},
   "outputs": [],
   "source": [
    "style = [\n",
    "    {\n",
    "        \"selector\": 'node',\n",
    "        \"style\": {\n",
    "            'background-color': 'data(color)',\n",
    "            'border-color': 'data(borderColor)',\n",
    "            'border-width': 'data(borderWidth)',\n",
    "            'label': 'data(label)',\n",
    "            \"text-valign\": \"center\",\n",
    "            \"text-halign\": \"center\",\n",
    "            'width': 'mapData(node_type, 0, 1, 70, 150)',\n",
    "            'height': 'mapData(node_type, 0, 1, 70, 150)',\n",
    "        }\n",
    "    },\n",
    "    {\n",
    "        \"selector\": 'edge',\n",
    "        \"style\": {\n",
    "            \"curve-style\": \"bezier\",\n",
    "            'line-color': 'data(lineColor)',\n",
    "            'width': '3',\n",
    "            \"text-rotation\": \"autorotate\",\n",
    "            \"text-margin-x\": 0,\n",
    "            \"text-margin-y\": 0,\n",
    "            'font-size': '12px',\n",
    "            'target-arrow-shape': 'data(directed)',\n",
    "            'target-endpoint': 'outside-to-node',\n",
    "            'source-endpoint': 'outside-to-node',\n",
    "            'target-arrow-color': 'data(lineColor)'\n",
    "        }\n",
    "    }\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31",
   "metadata": {},
   "outputs": [],
   "source": [
    "cytoscapeobj = ipc.CytoscapeWidget() \n",
    "cytoscapeobj.graph.add_graph_from_json(d) \n",
    "cytoscapeobj.set_style(style) \n",
    "cytoscapeobj.set_layout(name=sn_layout, animate=True) \n",
    "display(Markdown(f'## **{desc}**'))\n",
    "display(cytoscapeobj)\n",
    "\n",
    "display(Markdown('**Figure 3.** *Subnetwork of enriched TFs. The subnetwork displays each of the enriched TFs for the input gene set, connected with edges from the ChEA-KG background GRN.*'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
